doc/retro.1
.TH retro 1 "February 2012" "" "Retro Language"

.SH NAME
.P
retro - a stack-based language

.SH SYNOPSIS
.P
retro [options]

.SH DESCRIPTION
.P
Retro is a concatenative, stack-based language with roots in Forth.

.P
.B
--image
.I
filename
.RS
Use
.I
filename
as the image file to load/save
.RE

.P
.B
--shrink
.RS
Only save the used portion of the image. This reduces file size significantly.
.RE

.P
.B
--with
.I
filename
.RS
Add
.I
filename
to the input stack
.RE

.P
.B
--stats
.RS
Display statistics on opcodes processed upon exit
.RE

.SH FINDING THE IMAGE
.P
If you do not specify an image using
.B
--with
Retro will try to load a file named
.I
retroImage
from the current working directory. If this is not found, Retro
checks the
.B
RETROIMAGE
environment variable for the image to use. If this is not found,
an error will be reported.

.SH REPORTING BUGS
.P
Send bug reports to Charles Childers <crc@retroforth.org>

.SH COPYRIGHT
.P
Copyright (c) 2008 - 2012,
Charles Childers, Luke Parrish, Marc Simpson, Jay Skeer,
JGL, Marc Simpson, Greg Copeland, Aleksej Saushev,
Foucist, Erturk Kocalar, Kenneth Keating

.P
Permission to use, copy, modify, and/or distribute this software for
any purpose with or without fee is hereby granted, provided that the
above copyright notice and this permission notice appear in all
copies.

.P
THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL
WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE
AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL
DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR
PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER
TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.


RELEASE_NOTES.rst
=============
Release Notes
=============

----------
Retro 11.5
----------

core language / standard image
==============================

  - metacompiler changes

    - docstring support
    - create headers during metacompilation

  - clean ups and refactorings

    - increased use of & prefix over [ ]
    - comments on quotation compiler added (tangentstorm)

  - new functionality

    - added yield
    - added d->doc
    - added describe
    - added :needs
    - added libraries
    - added parable style conditionals (now default)
    - exposed getc:unfiltered
    - keymap support

  - moved functions to internal' vocabulary

    - quote
    - string

  - docstrings now part of standard image

    - make "smallimage" to get one without them


libraries
=========

  - dissect' now tries to display function names for calls
  - added havn hashes from Mat2 to hash'


examples
========

  - autopsy

    - added words:vectored
    - added words:byClass
    - use latest dissect'

  - block editor

    - fixed a few issues causing corruption (thanks to docl finding the bugs)

  - assembler


documentation
=============

  - various fixes


vm
==

  - improved python implementation from tangentstorm


----------
Retro 11.4
----------

Compatibility Issues
====================

This release brings several changes that may require small alterations to
existing source code.

First, the **files'** vocabulary is now in the library and not the core
language. For most users, adding a single line to the start of their sources
will be sufficient.

::

  needs files'

The second change is more significant. We have moved **pow**, **abs**, **min**,
**max**, and **random** to the **math'** library. If you have source using these,
you need to add the following to your sources:

::

  needs math'
  with  math'

If you don't want to do **with math'**, you can add a **^math'** prefix to each
use of these functions and just use the **needs math'** line.

Third, hidden headers are no longer preserved. There is a new function, **HEADERS**,
which returns the maximum number of hidden headers. Exceeding this will cause
problems, so it can be revectored to increase (or decrease) as needed.

Fourth, the **tib**, temporary string buffers, and hidden headers are now moved
to the end of physical memory. Use the **introspection'** library to determine
the size and actual locations of these buffers, and avoid overflowing into them.
E.g., in **casket'**, we use **^introspection'startOfBuffers** to identify the
start of the buffers and dynamically shift the casket buffers to a safe location.

Fifth, string constants are now created using **string:** instead of **string**.
This requires a quick search/replace to fix.


core language / standard image
==============================

  - clean ups and refactorings

    - **{{** and **}}**
    - **:is** and **:devector**
    - **each@**
    - **<puts>**
    - **toNumber**

  - removals (now in library)

    - **pow**
    - **abs**
    - **min**
    - **max**
    - **random**
    - **files'**

  - new memory layout (buffers at physical end of memory)
  - hidden headers now disappear when global scope resumes
  - helper function **string** for strings


libraries
=========

  - **d'** renamed to **decimal'** (resolve naming conflict)
  - add **queries'**
  - add **dump'**
  - add **fixed'**
  - add **double'**
  - add **unsigned'**
  - add **introspection'**
  - add **files'** (used to be in core image)
  - update **math'**

    - add **pow** (used to be in core image)
    - add **abs** (used to be in core image)
    - add **min** (used to be in core image)
    - add **max** (used to be in core image)
    - add **random** (used to be in core image)

  - updated **console'** to use new console device with VT100 fallback


examples
========

  - cleanups to make better use of higher-level language features

vm
==

  - add retro-curses
  - C# is now feature complete
  - added support for VM-level console device
  - new java implementation from Mike Andrews

website
=======

  - fix broken links





----------
Retro 11.3
----------


Compatibility Issues
====================

This release of Retro brings a number of changes that may require small
alterations to existing sources. Specifically:

  - maximum string length is now 256 cells by default
  - hidden functions are no longer revectorable

The first is not a big deal. It's now possible to alter both the maximum
string length and the number of temporary buffers. To restore a 512 cell
size as in 11.2:

::

  here 512 allot constant <512-TIB>
  [ <512-TIB> ] is tib
  [ 512 ] is STRING-LENGTH

The second change shouldn't be too critical. Basically it means that this
will no longer work:

::

  {{
    : foo  ( - ) 50 ;
  ---reveal---
    : bar1  foo foo + ;
    [ 100 ] is foo
  }}

Any private definitions are no longer revectorable. If you have no
**---reveal---** in a namespace, all functions are non-revectorable. This
won't break anything in the standard language or libraries, but could
be a problem if you rely on the old behavior in your code.


core language / standard image
==============================

  - new method of implementing quotes
  - maximum string length can be altered now
  - number of string buffers can be altered now

vm/complete/retro.c
/* Ngaro VM ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   Copyright (c) 2008 - 2011, Charles Childers
   Copyright (c) 2009 - 2010, Luke Parrish
   Copyright (c) 2010,        Marc Simpson
   Copyright (c) 2010,        Jay Skeer
   Copyright (c) 2011,        Kenneth Keating
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ */
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <unistd.h>
#include <string.h>
#include <termios.h>
#include <sys/ioctl.h>
/* ATH */
#include <sys/stat.h>
#include <errno.h>

/* Configuration ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

                +---------+---------+---------+
                | 16 bit  | 32 bit  | 64 bit  |
   +------------+---------+---------+---------+
   | IMAGE_SIZE | 32000   | 1000000 | 1000000 |
   +------------+---------+---------+---------+
   | CELL       | int16_t | int32_t | int64_t |
   +------------+---------+---------+---------+

   If memory is tight, cut the MAX_FILE_NAME and MAX_REQUEST_LENGTH.

   You can also cut the ADDRESSES stack size down, but if you have
   heavy nesting or recursion this may cause problems. If you do modify
   it and experience odd problems, try raising it a bit higher.

   Use -DRX16 to select defaults for 16-bit, or -DRX64 to select the
   defaults for 64-bit. Without these, the compiler will generate a
   standard 32-bit VM.

   Use -DRXBE to enable the BE suffix for big endian images. This is
   only useful on big endian systems.
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ */
#define CELL            int32_t
#define IMAGE_SIZE      1000000
#define ADDRESSES          1024*1024
#define STACK_DEPTH         128
#define PORTS                12
#define MAX_FILE_NAME      1024
#define MAX_REQUEST_LENGTH 1024
#define MAX_OPEN_FILES        8
#define LOCAL                 "retroImage"
#define CELLSIZE             32

#ifdef RX64
#undef CELL
#undef CELLSIZE
#undef LOCAL
#define CELL     int64_t
#define CELLSIZE 64
#define LOCAL    "retroImage64"
#endif

#ifdef RX16
#undef CELL
#undef CELLSIZE
#undef LOCAL
#undef IMAGE_SIZE
#define CELL        int16_t
#define CELLSIZE    16
#define IMAGE_SIZE  32000
#define LOCAL       "retroImage16"
#endif

#ifdef RXBE
#define LOCAL_FNAME (LOCAL "BE")
#define VM_ENDIAN 1
#else
#define LOCAL_FNAME (LOCAL)
#define VM_ENDIAN 0
#endif


enum vm_opcode {VM_NOP, VM_LIT, VM_DUP, VM_DROP, VM_SWAP, VM_PUSH, VM_POP,
                VM_LOOP, VM_JUMP, VM_RETURN, VM_GT_JUMP, VM_LT_JUMP,
                VM_NE_JUMP,VM_EQ_JUMP, VM_FETCH, VM_STORE, VM_ADD,
                VM_SUB, VM_MUL, VM_DIVMOD, VM_AND, VM_OR, VM_XOR, VM_SHL,
                VM_SHR, VM_ZERO_EXIT, VM_INC, VM_DEC, VM_IN, VM_OUT,
                VM_WAIT };
#define NUM_OPS VM_WAIT + 1

typedef struct {
  CELL sp, rsp, ip;
  CELL data[STACK_DEPTH];
  CELL address[ADDRESSES];
  CELL ports[PORTS];
  FILE *files[MAX_OPEN_FILES];
  FILE *input[MAX_OPEN_FILES];
  CELL isp;
  CELL image[IMAGE_SIZE];
  CELL shrink;
  CELL filecells; // in image on disk
  int stats[NUM_OPS + 1];
  int max_sp, max_rsp;
  char filename[MAX_FILE_NAME];
  char request[MAX_REQUEST_LENGTH];
  struct termios new_termios, old_termios;
} VM;

/* Macros ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ */
#define IP   vm->ip
#define SP   vm->sp
#define RSP  vm->rsp
#define DROP { vm->data[SP] = 0; if (--SP < 0) IP = IMAGE_SIZE; }
#define TOS  vm->data[SP]
#define NOS  vm->data[SP-1]
#define TORS vm->address[RSP]

/* Helper Functions ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ */
void rxGetString(VM *vm, int starting)
{
  CELL i = 0;
  while(vm->image[starting] && i < MAX_REQUEST_LENGTH)
    vm->request[i++] = (char)vm->image[starting++];
  vm->request[i] = 0;
}

/* Console I/O Support ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ */
void rxWriteConsole(CELL c) {
  (c > 0) ? putchar((char)c) : printf("\033[2J\033[1;1H");
  /* Erase the previous character if c = backspace */
  if (c == 8) {
    putchar(32);
    putchar(8);
  }
}

CELL rxReadConsole(VM *vm) {
  CELL c;
  if ((c = getc(vm->input[vm->isp])) == EOF && vm->input[vm->isp] != stdin) {
    fclose(vm->input[vm->isp--]);
    c = 0;
  }
  if (c == EOF && vm->input[vm->isp] == stdin)
    exit(0);
  return c;
}

void rxIncludeFile(VM *vm, char *s) {
  FILE *file;
  if ((file = fopen(s, "r")))
    vm->input[++vm->isp] = file;
}

void rxPrepareInput(VM *vm) {
  vm->isp = 0;
  vm->input[vm->isp] = stdin;
}

void rxPrepareOutput(VM *vm) {
  tcgetattr(0, &vm->old_termios);
  vm->new_termios = vm->old_termios;
  vm->new_termios.c_iflag &= ~(BRKINT+ISTRIP+IXON+IXOFF);
  vm->new_termios.c_iflag |= (IGNBRK+IGNPAR);
  vm->new_termios.c_lflag &= ~(ICANON+ISIG+IEXTEN+ECHO);
  vm->new_termios.c_cc[VMIN] = 1;
  vm->new_termios.c_cc[VTIME] = 0;
  tcsetattr(0, TCSANOW, &vm->new_termios);
}

void rxRestoreIO(VM *vm) {
  tcsetattr(0, TCSANOW, &vm->old_termios);
}

/* File I/O Support ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ */
CELL rxGetFileHandle(VM *vm)
{
  CELL i;
  for(i = 1; i < MAX_OPEN_FILES; i++)
    if (vm->files[i] == 0)
      return i;
  return 0;
}

void rxAddInputSource(VM *vm) {
  CELL name = TOS; DROP;
  rxGetString(vm, name);
  rxIncludeFile(vm, vm->request);
}

CELL rxOpenFile(VM *vm) {
  CELL slot, mode, name;
  slot = rxGetFileHandle(vm);
  mode = TOS; DROP;
  name = TOS; DROP;
  rxGetString(vm, name);
  if (slot > 0)
  {
    if (mode == 0)  vm->files[slot] = fopen(vm->request, "r");
    if (mode == 1)  vm->files[slot] = fopen(vm->request, "w");
    if (mode == 2)  vm->files[slot] = fopen(vm->request, "a");
    if (mode == 3)  vm->files[slot] = fopen(vm->request, "r+");
  }
  if (vm->files[slot] == NULL)
  {
    vm->files[slot] = 0;
    slot = 0;
  }
  return slot;
}

CELL rxReadFile(VM *vm) {
  CELL c = fgetc(vm->files[TOS]); DROP;
  return (c == EOF) ? 0 : c;
}

CELL rxWriteFile(VM *vm) {
  CELL slot, c, r;
  slot = TOS; DROP;
  c = TOS; DROP;
  r = fputc(c, vm->files[slot]);
  return (r == EOF) ? 0 : 1;
}

CELL rxCloseFile(VM *vm) {
  fclose(vm->files[TOS]);
  vm->files[TOS] = 0;
  DROP;
  return 0;
}

CELL rxGetFilePosition(VM *vm) {
  CELL slot = TOS; DROP;
  return (CELL) ftell(vm->files[slot]);
}

CELL rxSetFilePosition(VM *vm) {
  CELL slot, pos, r;
  slot = TOS; DROP;
  pos  = TOS; DROP;
  r = fseek(vm->files[slot], pos, SEEK_SET);
  return r;
}

CELL rxGetFileSize(VM *vm) {
  CELL slot, current, r, size;
  slot = TOS; DROP;
  current = ftell(vm->files[slot]);
  r = fseek(vm->files[slot], 0, SEEK_END);
  size = ftell(vm->files[slot]);
  fseek(vm->files[slot], current, SEEK_SET);
  return (r == 0) ? size : 0;
}

CELL rxDeleteFile(VM *vm) {
  CELL name = TOS; DROP;
  rxGetString(vm, name);
  return (unlink(vm->request) == 0) ? -1 : 0;
}

CELL rxLoadImage(VM *vm, char *image) {
  FILE *fp;
  CELL x = 0;

  if ((fp = fopen(image, "rb")) != NULL) {
    x = fread(&vm->image, sizeof(CELL), IMAGE_SIZE, fp);
    fclose(fp);
  }
  else {
    printf("Unable to find the retroImage!\n");
    exit(1);
  }
  vm->filecells = x;
  return x;
}

CELL rxSaveImage(VM *vm, char *image) {
  FILE *fp;
  CELL x = 0;

  if ((fp = fopen(image, "wb")) == NULL)
  {
    printf("Unable to save the retroImage!\n");
    rxRestoreIO(vm);
    exit(2);
  }

  if (vm->shrink == 0)
    x = fwrite(&vm->image, sizeof(CELL), IMAGE_SIZE, fp);
  else
    x = fwrite(&vm->image, sizeof(CELL), vm->image[3], fp);
  fclose(fp);

  return x;
}

/* Environment Query ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ */
void rxQueryEnvironment(VM *vm) {
  CELL req, dest;
  char *r;
  req = TOS;  DROP;
  dest = TOS; DROP;

  rxGetString(vm, req);
  r = getenv(vm->request);

  if (r != 0)
    while (*r != '\0')
    {
      vm->image[dest] = *r;
      dest++;
      vm->image[dest] = 0;
      r++;
    }
  else
    vm->image[dest] = 0;
}

/* Device I/O Handler ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ */
void rxDeviceHandler(VM *vm) {
  struct winsize w;
  if (vm->ports[0] != 1) {
    /* Input */
    if (vm->ports[0] == 0 && vm->ports[1] == 1) {
      vm->ports[1] = rxReadConsole(vm);
      vm->ports[0] = 1;
    }

    /* Output (character generator) */
    if (vm->ports[2] == 1) {
      rxWriteConsole(TOS); DROP
      vm->ports[2] = 0;
      vm->ports[0] = 1;
    }

    /* File IO and Image Saving */
    if (vm->ports[4] != 0) {
      vm->ports[0] = 1;
      switch (vm->ports[4]) {
        case  1: rxSaveImage(vm, vm->filename);
                 vm->ports[4] = 0;
                 break;
        case  2: rxAddInputSource(vm);
                 vm->ports[4] = 0;
                 break;
        case -1: vm->ports[4] = rxOpenFile(vm);
                 break;
        case -2: vm->ports[4] = rxReadFile(vm);
                 break;
        case -3: vm->ports[4] = rxWriteFile(vm);
                 break;
        case -4: vm->ports[4] = rxCloseFile(vm);
                 break;
        case -5: vm->ports[4] = rxGetFilePosition(vm);
                 break;
        case -6: vm->ports[4] = rxSetFilePosition(vm);
                 break;
        case -7: vm->ports[4] = rxGetFileSize(vm);
                 break;
        case -8: vm->ports[4] = rxDeleteFile(vm);
                 break;
        default: vm->ports[4] = 0;
      }
    }

    /* Capabilities */
    if (vm->ports[5] != 0) {
      vm->ports[0] = 1;
      switch(vm->ports[5]) {
        case -1:  vm->ports[5] = IMAGE_SIZE;
                  break;
        case -2:  vm->ports[5] = 0;
                  break;
        case -3:  vm->ports[5] = 0;
                  break;
        case -4:  vm->ports[5] = 0;
                  break;
        case -5:  vm->ports[5] = SP;
                  break;
        case -6:  vm->ports[5] = RSP;
                  break;
        case -7:  vm->ports[5] = 0;
                  break;
        case -8:  vm->ports[5] = time(NULL);
                  break;
        case -9:  vm->ports[5] = 0;
                  IP = IMAGE_SIZE;
                  break;
        case -10: vm->ports[5] = 0;
                  rxQueryEnvironment(vm);
                  break;
        case -11: ioctl(0, TIOCGWINSZ, &w);
                  vm->ports[5] = w.ws_col;
                  break;
        case -12: ioctl(0, TIOCGWINSZ, &w);
                  vm->ports[5] = w.ws_row;
                  break;
        case -13: vm->ports[5] = CELLSIZE;
                  break;
        case -14: vm->ports[5] = VM_ENDIAN;
                  break;
        case -15: vm->ports[5] = -1;
                  break;
        case -16: vm->ports[5] = STACK_DEPTH;
                  break;
        case -17: vm->ports[5] = ADDRESSES;
                  break;
        default:  vm->ports[5] = 0;
      }
    }

    if (vm->ports[8] != 0) {
      switch (vm->ports[8]) {
        case 1: vm->ports[8] = 0;
                printf("\e[%d;%dH", NOS, TOS);
                DROP; DROP;
                break;
        case 2: vm->ports[8] = 0;
                printf("\e[3%dm", TOS);
                DROP;
                break;
        case 3: vm->ports[8] = 0;
                printf("\e[4%dm", TOS);
                DROP;
                break;
        case 4: vm->ports[8] = 0;
                break;
        default: vm->ports[8] = 0;
      }
    }
  }
}

/* The VM ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ */
void rxProcessOpcode(VM *vm) {
  CELL a, b, opcode;
  opcode = vm->image[IP];

  if (opcode > NUM_OPS)
    vm->stats[NUM_OPS]++;
  else
    vm->stats[opcode]++;

  switch(opcode) {
    case VM_NOP:
         break;
    case VM_LIT:
         SP++;
         IP++;
         TOS = vm->image[IP];
         if (vm->max_sp < SP)
           vm->max_sp = SP;
         break;
    case VM_DUP:
         SP++;
         vm->data[SP] = NOS;
         if (vm->max_sp < SP)
           vm->max_sp = SP;
         break;
    case VM_DROP:
         DROP
         break;
    case VM_SWAP:
         a = TOS;
         TOS = NOS;
         NOS = a;
         break;
    case VM_PUSH:
         RSP++;
         TORS = TOS;
         DROP
         if (vm->max_rsp < RSP)
           vm->max_rsp = RSP;
         break;
    case VM_POP:
         SP++;
         TOS = TORS;
         RSP--;
         break;
    case VM_LOOP:
         TOS--;
           IP++;
         if (TOS != 0 && TOS > -1)
           IP = vm->image[IP] - 1;
         else
           DROP;
         break;
    case VM_JUMP:
         IP++;
         IP = vm->image[IP] - 1;
         if (vm->image[IP+1] == 0)
           IP++;
         if (vm->image[IP+1] == 0)
           IP++;
         break;
    case VM_RETURN:
         IP = TORS;
         RSP--;
         if (vm->image[IP+1] == 0)
           IP++;
         if (vm->image[IP+1] == 0)
           IP++;
         break;
    case VM_GT_JUMP:
         IP++;
         if(NOS > TOS)
           IP = vm->image[IP] - 1;
         DROP DROP
         break;
    case VM_LT_JUMP:
         IP++;
         if(NOS < TOS)
           IP = vm->image[IP] - 1;
         DROP DROP
         break;
    case VM_NE_JUMP:
         IP++;
         if(TOS != NOS)
           IP = vm->image[IP] - 1;
         DROP DROP
         break;
    case VM_EQ_JUMP:
         IP++;
         if(TOS == NOS)
           IP = vm->image[IP] - 1;
         DROP DROP
         break;
    case VM_FETCH:
         TOS = vm->image[TOS];
         break;
    case VM_STORE:
         vm->image[TOS] = NOS;
         DROP DROP
         break;
    case VM_ADD:
         NOS += TOS;
         DROP
         break;
    case VM_SUB:
         NOS -= TOS;
         DROP
         break;
    case VM_MUL:
         NOS *= TOS;
         DROP
         break;
    case VM_DIVMOD:
         a = TOS;
         b = NOS;
         TOS = b / a;
         NOS = b % a;
         break;
    case VM_AND:
         a = TOS;
         b = NOS;
         DROP
         TOS = a & b;
         break;
    case VM_OR:
         a = TOS;
         b = NOS;
         DROP
         TOS = a | b;
         break;
    case VM_XOR:
         a = TOS;
         b = NOS;
         DROP
         TOS = a ^ b;
         break;
    case VM_SHL:
         a = TOS;
         b = NOS;
         DROP
         TOS = b << a;
         break;
    case VM_SHR:
         a = TOS;
         DROP
         TOS >>= a;
         break;
    case VM_ZERO_EXIT:
         if (TOS == 0) {
           DROP
           IP = TORS;
           RSP--;
         }
         break;
    case VM_INC:
         TOS += 1;
         break;
    case VM_DEC:
         TOS -= 1;
         break;
    case VM_IN:
         a = TOS;
         TOS = vm->ports[a];
         vm->ports[a] = 0;
         break;
    case VM_OUT:
         vm->ports[TOS] = NOS;
         DROP DROP
         break;
    case VM_WAIT:
         rxDeviceHandler(vm);
         break;
    default:
         RSP++;
         TORS = IP;
         IP = vm->image[IP] - 1;

         if (vm->image[IP+1] == 0)
           IP++;
         if (vm->image[IP+1] == 0)
           IP++;

         if (vm->max_rsp < RSP)
           vm->max_rsp = RSP;
         break;
  }
  vm->ports[3] = 1;
}

/* Stats ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ */
void rxDisplayStats(VM *vm)
{
  int s, i;

  printf("Runtime Statistics\n");
  printf("NOP:     %d\n", vm->stats[VM_NOP]);
  printf("LIT:     %d\n", vm->stats[VM_LIT]);
  printf("DUP:     %d\n", vm->stats[VM_DUP]);
  printf("DROP:    %d\n", vm->stats[VM_DROP]);
  printf("SWAP:    %d\n", vm->stats[VM_SWAP]);
  printf("PUSH:    %d\n", vm->stats[VM_PUSH]);
  printf("POP:     %d\n", vm->stats[VM_POP]);
  printf("LOOP:    %d\n", vm->stats[VM_LOOP]);
  printf("JUMP:    %d\n", vm->stats[VM_JUMP]);
  printf("RETURN:  %d\n", vm->stats[VM_RETURN]);
  printf(">JUMP:   %d\n", vm->stats[VM_GT_JUMP]);
  printf("<JUMP:   %d\n", vm->stats[VM_LT_JUMP]);
  printf("!JUMP:   %d\n", vm->stats[VM_NE_JUMP]);
  printf("=JUMP:   %d\n", vm->stats[VM_EQ_JUMP]);
  printf("FETCH:   %d\n", vm->stats[VM_FETCH]);
  printf("STORE:   %d\n", vm->stats[VM_STORE]);
  printf("ADD:     %d\n", vm->stats[VM_ADD]);
  printf("SUB:     %d\n", vm->stats[VM_SUB]);
  printf("MUL:     %d\n", vm->stats[VM_MUL]);
  printf("DIVMOD:  %d\n", vm->stats[VM_DIVMOD]);
  printf("AND:     %d\n", vm->stats[VM_AND]);
  printf("OR:      %d\n", vm->stats[VM_OR]);
  printf("XOR:     %d\n", vm->stats[VM_XOR]);
  printf("SHL:     %d\n", vm->stats[VM_SHL]);
  printf("SHR:     %d\n", vm->stats[VM_SHR]);
  printf("0;:      %d\n", vm->stats[VM_ZERO_EXIT]);
  printf("INC:     %d\n", vm->stats[VM_INC]);
  printf("DEC:     %d\n", vm->stats[VM_DEC]);
  printf("IN:      %d\n", vm->stats[VM_IN]);
  printf("OUT:     %d\n", vm->stats[VM_OUT]);
  printf("WAIT:    %d\n", vm->stats[VM_WAIT]);
  printf("CALL:    %d\n", vm->stats[NUM_OPS]);
  printf("Max SP:  %d\n", vm->max_sp);
  printf("Max RSP: %d\n", vm->max_rsp);

  for (s = i = 0; s < NUM_OPS; s++)
    i += vm->stats[s];
  printf("Total opcodes processed: %d\n", i);
}

/* Dump ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ */

// see /test/ngaro/ngarotest.py

#define FILSEP printf( "\x1C" )
#define GRPSEP printf( "\x1D" )

// print all cells in an array, separated by spaces
// stacks use base 1 indexing, but ram uses base 0
// so we have to compensate for that
void dumparray( CELL arr[], int base, int last ) {
  int i, count;
  count = last - base + 1;
  // printf( "\n\nlast: %i   data[0:2] = [ %i %i ]\n", last, arr[ 0 ], arr[ 1 ] );
  // printf( "count = %d\n\n", count );
  if ( count > 0 ) {
    if ( count > 1 )
      for ( i = 0; i < count - 1; ++i )
	printf( "%i ", arr[ base + i ]);
    printf( "%i", arr[ last ]);
  }
}

void rxDump(VM *vm) {
  FILSEP;
  dumparray( vm->data, 1, SP );
  GRPSEP;
  dumparray( vm->address, 1, RSP );
  GRPSEP;
  dumparray( vm->image, 0, vm->filecells - 1 );
}


/* Main ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ */
int main(int argc, char **argv) {
  VM *vm;
  int i, wantsStats, dumpAfter;

  /* ATH */
  char *env;
  struct stat sts;

  wantsStats = 0;
  vm = calloc(sizeof(VM), sizeof(char));
  strcpy(vm->filename, LOCAL_FNAME);

  rxPrepareInput(vm);

  /* Parse the command line arguments */
  for (i = 1; i < argc; i++) {
    if (strcmp(argv[i], "--with") == 0)
      rxIncludeFile(vm, argv[++i]);
    if (strcmp(argv[i], "--image") == 0)
      strcpy(vm->filename, argv[++i]);
    if (strcmp(argv[i], "--shrink") == 0)
      vm->shrink = 1;
    if (strcmp(argv[i], "--stats") == 0)
      wantsStats = 1;
    if (strcmp(argv[i], "--dump") == 0)
      dumpAfter = 1;
    if (strcmp(argv[i], "--help") == 0)
    {
      printf("--with filename    Add filename to the input stack\n");
      printf("--image filename   Use filename as the image to load\n");
      printf("--shrink           When saving, don't save unused cells\n");
      printf("--stats            Display opcode usage and stack summaries upon exit\n");
      printf("--dump             Dump stacks and image upon exit, for ngarotest.py\n");
      printf("--help             Display this text\n");
      exit(1);
    }
  }

  /* ATH - 26 January 2012
   *
   * Check for the existence of the file name held in vm->filename
   * If it does not exist, read the env variable RETROIMAGE and check for
   * the existence of that file. If that file does not exist,
   * exit with an error.
   *
   */

  if ( ( stat( vm->filename, &sts) == -1 ) && errno == ENOENT ) {
      // File doesn't exist, get the environment variable.
      env = (char *)getenv("RETROIMAGE");
      if( !env ) {
        fprintf(stderr,"No image file and environment variable RETROIMAGE not set.\n");
        exit(1);
      } else {
          strncpy(vm->filename, env,sizeof(vm->filename));
          fprintf(stderr,"Loading image from %s\n", env);
      }
  }
  if (rxLoadImage(vm, vm->filename) == 0) {
    printf("Sorry, unable to find %s\n", vm->filename);
    free(vm);
    exit(1);
  }

  rxPrepareOutput(vm);
  for (IP = 0; IP < IMAGE_SIZE; IP++)
    rxProcessOpcode(vm);
  rxRestoreIO(vm);

  if (wantsStats == 1)
    rxDisplayStats(vm);
  if (dumpAfter == 1)
    rxDump( vm );

  free(vm);
  return 0;
}
